lib/Compiler.js

```diff
@@ -25,6 +24,7 @@ const RequestShortener = require("./RequestShortener");
 const ResolverFactory = require("./ResolverFactory");
 const Stats = require("./Stats");
 const Watching = require("./Watching");
+const { join, dirname, mkdirp } = require("./util/fs");
```

Comment:
Can we export these utils as part of public api? It is very useful for `webpack-dev-middleware`

Reply:
I would better avoid exposing them as public API. But it should be fine to copy them to webpack-dev-middleware instead.

Reply:
Maybe we can use `require('webpack/util/fs')`? We already do this for `ModuleFilenameHelpers` https://github.com/webpack-contrib/terser-webpack-plugin/blob/master/src/index.js#L10